From b192b63d96cb9b08da93aa0d2080754ff6a3a61d Mon Sep 17 00:00:00 2001
From: P Praneesh <quic_ppranees@quicinc.com>
Date: Thu, 25 Aug 2022 14:40:02 +0530
Subject: [PATCH] ath12k: fill peer meta data during reo_reinject

Firmware expects physical address in the reo entrance ring when MLO is
enabled and peer meta data in the queue_addr_lo for the non MLO cases.
Current implementation fills the physical address for the non MLO cases
and firmware decodes it differently which leads to SOC CSR region access
which is not related to firmware/MAC operation.

Signed-off-by: P Praneesh <quic_ppranees@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp_rx.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/dp_rx.c b/drivers/net/wireless/ath/ath12k/dp_rx.c
index c4ef2e4..a8a9bce 100644
--- a/drivers/net/wireless/ath/ath12k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_rx.c
@@ -3437,10 +3437,12 @@ static int ath12k_dp_rx_h_defrag_reo_reinject(struct ath12k *ar, struct dp_rx_ti
 	reo_ent_ring->rx_mpdu_info.peer_meta_data =
 		reo_dest_ring->rx_mpdu_info.peer_meta_data;
 
-	reo_ent_ring->queue_addr_lo = lower_32_bits(rx_tid->paddr);
-	reo_ent_ring->info0 = FIELD_PREP(HAL_REO_ENTR_RING_INFO0_QUEUE_ADDR_HI,
-					 upper_32_bits(rx_tid->paddr)) |
-			      FIELD_PREP(HAL_REO_ENTR_RING_INFO0_DEST_IND, dst_ind);
+	/* Firmware expects physical address to be filled in queue_addr_lo in
+	 * the MLO scenario and in case of non MLO peer meta data needs to be
+	 * filled.
+	 */
+	reo_ent_ring->queue_addr_lo = reo_dest_ring->rx_mpdu_info.peer_meta_data;
+	reo_ent_ring->info0 = FIELD_PREP(HAL_REO_ENTR_RING_INFO0_DEST_IND, dst_ind);
 
 	reo_ent_ring->info1 = FIELD_PREP(HAL_REO_ENTR_RING_INFO1_MPDU_SEQ_NUM,
 					 rx_tid->cur_sn);
-- 
2.7.4

